src/zenml/zen_server/routers/plugin_endpoints.py

```diff
@@ -37,82 +44,61 @@
 
 @plugin_router.get(
     "",
-    response_model=List[str],
+    response_model=Page[BasePluginFlavorResponse],
     responses={401: error_response, 404: error_response, 422: error_response},
 )
 @handle_exceptions
-def list_plugin_flavors(
-    _: AuthContext = Security(authorize),
-) -> List[str]:
-    """Returns all event flavors.
-
-    Returns:
-        All flavors.
-    """
-    return plugin_flavor_registry.available_flavors
-
-
-# -------------------- Plugin Types --------------------
-
-
-@plugin_router.get(
-    "/{flavor_name}",
-    response_model=List[str],
-    responses={401: error_response, 404: error_response, 422: error_response},
-)
-@handle_exceptions
-def list_plugin_types(
-    flavor_name: str,
+def list_flavors(
+    plugin_type: PluginType,
+    plugin_subtype: PluginSubType,
+    page: int = PAGINATION_STARTING_PAGE,
+    size: int = PAGE_SIZE_DEFAULT,
     _: AuthContext = Security(authorize),
-) -> List[str]:
+) -> Page[BasePluginFlavorResponse]:
     """Returns all event flavors.
 
     Returns:
         All flavors.
     """
-    return plugin_flavor_registry.get_available_types_for_flavor(
-        flavor_name=flavor_name
+    flavors = (
+        plugin_flavor_registry.list_available_flavors_for_type_and_subtype(
+            _type=plugin_type,
+            sub_type=plugin_subtype,
+        )
     )
-
-
-# -------------------- Plugin SubTypes --------------------
-
-
-@plugin_router.get(
-    "/{flavor_name}/{plugin_type}",
-    response_model=List[str],
-    responses={401: error_response, 404: error_response, 422: error_response},
-)
-@handle_exceptions
-def list_plugin_subtypes(
-    flavor_name: str,
-    plugin_type: PluginType,
-    _: AuthContext = Security(authorize),
-) -> List[str]:
-    """Returns all subtypes.
-
-    Returns:
-        All subtypes for the given flavor and type.
-    """
-    return plugin_flavor_registry.get_available_subtypes_for_flavor_and_type(
-        flavor=flavor_name, _type=plugin_type
+    total = len(flavors)
+    total_pages = total / size
+    start = (page - 1) * size
+    end = start + size
+
+    page_items = [
+        flavor.get_plugin_flavor_response_model() for flavor in flavors
+    ][start:end]
+
+    return_page = Page(
+        index=page,
+        max_size=size,
+        total_pages=total_pages,
+        total=total,
+        items=page_items,
```

Comment:
With most (maybe all ?) of the other "list" endpoints, we usually just call the sql zen store to handle everything related to filtering and pagination. Do you think we could also to that here and move all this logic into the plugin registry ?

Reply:
Exactly what I wanted to do, was holding myself back though to optimize for speed. Your comment makes me itch to do it anyways now :D

Reply:
I actually wanted to even have a generic way to paginate in_memory list on the InMemoryFilterModel, but I think thats for the future